File: game-model-v2/src/homemade/game/model/CellState.java
package homemade.game.model;

import java.util.EnumSet;
import java.util.Set;

/**
 * Extensions of this class must be immutable
 */
public abstract class CellState {
    private static Set<Cell> cellFreeToMove = EnumSet.of(Cell.EMPTY, Cell.MARKED_FOR_SPAWN);
    private static Set<Cell> blocks = EnumSet.of(Cell.OCCUPIED, Cell.DEAD_BLOCK);

    public final static int UNDEFINED_VALUE = -1;
    public final static ComboEffect UNDEFINED_COMBO_EFFECT = null;
    public final static boolean UNDEFINED_PROPERTY = false;

    private final Cell cellType;

    protected CellState(Cell cellType) {
        this.cellType = cellType;
    }

    public int value() {
        return UNDEFINED_VALUE;
    }

    public ComboEffect effect() {
        return UNDEFINED_COMBO_EFFECT;
    }

    public boolean isMovableBlock() {
        return UNDEFINED_PROPERTY;
    }

    /**
     * I think we need this method for view
     */
    public final Cell type() {
        return cellType;
    }

    public final boolean isAliveBlock() { //TODO: find a better method name
        return cellType == Cell.OCCUPIED;
    }

    public final boolean isAnyBlock() {
        return blocks.contains(cellType);
    }

    public final boolean isFreeForMove() {
        return cellFreeToMove.contains(cellType);
    }

    public final boolean isFreeForSpawn() {
        return cellType == Cell.EMPTY;
    }
}

File: game-model-v2/src/homemade/game/model/CellState.kt
package homemade.game.model

import homemade.game.model.combo.ComboEffect
import java.util.*

/**
 * Extensions of this class must be immutable
 */
abstract class CellState protected constructor(private val cellType: Cell) {
    open fun value(): Int {
        return UNDEFINED_VALUE
    }

    open fun effect(): ComboEffect? {
        return ComboEffect.UNDEFINED_COMBO_EFFECT
    }
    //TODO somethings wrong with isMovableBlock and isAliveBlock

    fun type(): Cell { //TODO why extra method
        return cellType
    }

    val isAliveBlock: Boolean
        get() = cellType == Cell.OCCUPIED
    val isAnyBlock: Boolean
        get() = blocks.contains(cellType)
    val isFreeForMove: Boolean
        get() = cellFreeToMove.contains(cellType)
    val isFreeForSpawn: Boolean
        get() = cellType == Cell.EMPTY
    open val isMovableBlock: Boolean
        get() = false

    companion object {
        private val cellFreeToMove: Set<Cell> = EnumSet.of(Cell.EMPTY, Cell.MARKED_FOR_SPAWN)
        private val blocks: Set<Cell> = EnumSet.of(Cell.OCCUPIED, Cell.DEAD_BLOCK)
        const val UNDEFINED_VALUE = -1
    }
}
